bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/AbstractExpression.java

```diff
@@ -207,22 +207,51 @@ protected void prune() {
 
     @Override
     public Date getTimeAfter(Date afterTime) {
+
+        Date currentStartDate = getStartDate();
+
+        if (hasFloatingStartDate()) {
+            try {
+                clearCandidates();
+                setStartDate(afterTime);
+            } catch (IllegalArgumentException | ParseException e) {
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
+            }
+        }
+
         if (getCandidates().isEmpty()) {
             try {
                 setStartDate(afterTime);
-                parseExpression(expression);
             } catch (ParseException e) {
-                logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
             }
         }
 
         if (!getCandidates().isEmpty()) {
-
-            Collections.sort(getCandidates());
-
-            for (Date candidate : getCandidates()) {
-                if (candidate.after(afterTime)) {
-                    return candidate;
+            if (getCandidates().size() == 1) {
+                return getCandidates().get(0);
+            } else {
+                while (getCandidates().size() > 1) {
```

Comment:
Hm, I do not get it.

Isn't at the end in the while loop by using `clearCandidates()` and `setStartDate(newStartDate)` the candidate set new calculated?

Then the next iteration starts with the check for &gt 1...

```java
// First loop iteration
// Let's assume: getCandidates().size() == 5 (at least something > 1)

// The condition "getCandidates().size() > 1" is checked
// All fine


    Collections.sort(getCandidates());
    Date newStartDate = null;
    try {
        for (Date candidate : getCandidates()) {
            newStartDate = candidate;
            if (candidate.after(afterTime)) {
                setStartDate(currentStartDate);
                return candidate;
            }
        }

// No of the current 5 candidates is after(afterTime)
// So we are here

// Clear the candidates and get new candidates by setting the start date
        clearCandidates();
        setStartDate(newStartDate);
// Let's assume the number of caluclated candidates 1: getCandidates().size() == 1

    } catch (IllegalArgumentException | ParseException e) {
        logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
    }

// Second iteration loop
// The condition "getCandidates().size() > 1" is checked
// The size is 1 so not greater then 1 and the loop is not entered.
```